stream/src/main/java/org/apache/apex/malhar/stream/api/Option.java

```diff
@@ -18,39 +18,87 @@
  */
 package org.apache.apex.malhar.stream.api;
 
-import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.apex.malhar.lib.window.ControlTuple;
 
 import com.datatorrent.api.Attribute;
+import com.datatorrent.api.Component;
+import com.datatorrent.api.Context;
 
 /**
- * Options for the operators in the dag
- *
- * @since 3.5.0
+ * Options for each transformation in the dag
  */
-@InterfaceStability.Evolving
 public interface Option
 {
+  abstract class WatermarkGenerator<T> implements Option, Component<Context.OperatorContext>
+  {
+
+    @Override
+    public void setup(Context.OperatorContext context)
+    {
+
+    }
+
+    @Override
+    public void teardown()
+    {
+
+    }
 
+    /**
+     * The current watermark generation will be called each endWindow to generate watermark
+     * @return null if you don't want to generate watermark for this window
+     */
+    public abstract ControlTuple.Watermark currentWatermark();
+
+    /**
+     * Generate the watermark from current tuple
+     * @param input
+     * @return null if you can't generate watermark from current tuple
+     */
+    public abstract ControlTuple.Watermark getWatermarkFromTuple(T input);
```

Comment:
Since we don't typically generate a new watermark tuple for each input tuple, how about making this return void and rename this to `process`?